File: python/problem_53.py
from typing import Optional


class StackQueue:
    def __init__(self):
        self._inbox = []
        self._outbox = []

    def enqueue(self, val: int) -> None:
        self._inbox.append(val)

    def dequeue(self) -> Optional[int]:
        if not self._outbox:
            while self._inbox:
                self._outbox.append(self._inbox.pop())
        return self._outbox.pop() if self._outbox else None


def main():
    print("Problem 53:")
    q = StackQueue()
    q.enqueue(1); q.enqueue(2); q.enqueue(3)
    assert q.dequeue() == 1
    assert q.dequeue() == 2
    q.enqueue(4)
    assert q.dequeue() == 3
    assert q.dequeue() == 4
    assert q.dequeue() is None
    print("All Tests Passed")

File: c/problem_53.c
#include "problem_53.h"
#include <stdio.h>
#include <assert.h>

void sq_init(StackQueue *q) { q->in_sz = q->out_sz = 0; }

void sq_enqueue(StackQueue *q, int v) { q->inbox[q->in_sz++] = v; }

int sq_dequeue(StackQueue *q) {
    if (q->out_sz == 0) {
        while (q->in_sz > 0) q->outbox[q->out_sz++] = q->inbox[--q->in_sz];
    }
    if (q->out_sz == 0) return -1;
    return q->outbox[--q->out_sz];
}

void problem53(void) {
    printf("Problem 53:\n");
    StackQueue q; sq_init(&q);
    sq_enqueue(&q,1); sq_enqueue(&q,2); sq_enqueue(&q,3);
    assert(sq_dequeue(&q)==1);
    assert(sq_dequeue(&q)==2);
    sq_enqueue(&q,4);
    assert(sq_dequeue(&q)==3);
    assert(sq_dequeue(&q)==4);
    assert(sq_dequeue(&q)==-1);
    printf("All Tests Passed\n");
}
